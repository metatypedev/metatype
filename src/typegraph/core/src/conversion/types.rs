// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::wit::core::PolicySpec;
use common::typegraph::{Injection, PolicyIndices, TypeNode, TypeNodeBase};
use enum_dispatch::enum_dispatch;
use indexmap::IndexMap;
use std::rc::Rc;

use crate::errors::Result;
use crate::typegraph::TypegraphContext;
use crate::types::TypeId;

#[enum_dispatch]
pub trait TypeConversion {
    /// takes already converted runtime id
    fn convert(&self, ctx: &mut TypegraphContext, runtime_id: Option<u32>) -> Result<TypeNode>;
}

impl<T: TypeConversion> TypeConversion for Rc<T> {
    fn convert(&self, ctx: &mut TypegraphContext, runtime_id: Option<u32>) -> Result<TypeNode> {
        (**self).convert(ctx, runtime_id)
    }
}

pub struct BaseBuilderInit<'a, 'b> {
    pub ctx: &'a mut TypegraphContext,
    pub base_name: &'static str,
    pub type_id: TypeId,
    pub name: Option<String>,
    pub runtime_idx: u32,
    pub policies: &'b [PolicySpec],
    pub runtime_config: Option<&'b [(String, String)]>,
}

pub struct BaseBuilder<'a> {
    ctx: &'a mut TypegraphContext,
    name: String,
    runtime_idx: u32,
    policies: Vec<PolicyIndices>,
    runtime_config: Option<IndexMap<String, String>>,

    // optional features
    enumeration: Option<Vec<String>>,
    injection: Option<Injection>,
    as_id: bool,
}

impl<'a, 'b> BaseBuilderInit<'a, 'b> {
    pub fn init_builder(self) -> Result<BaseBuilder<'a>> {
        let policies = self.ctx.register_policy_chain(self.policies)?;

        let name = match self.name {
            Some(name) => name,
            None => format!("{}_{}_placeholder", self.base_name, self.type_id.0),
        };

        let runtime_config = self.runtime_config.map(|c| {
            c.iter()
                .map(|(k, v)| (k.to_string(), v.to_string()))
                .collect::<IndexMap<_, _>>()
        });

        Ok(BaseBuilder {
            ctx: self.ctx,
            name,
            runtime_idx: self.runtime_idx,
            policies,
            runtime_config,

            enumeration: None,
            injection: None,
            as_id: false,
        })
    }
}

impl<'a> BaseBuilder<'a> {
    pub fn build(self) -> Result<TypeNodeBase> {
        let config = self.runtime_config.map(|c| {
            c.iter()
                .map(|(k, v)| (k.to_string(), serde_json::from_str(v).unwrap()))
                .collect::<IndexMap<_, _>>()
        });

        Ok(TypeNodeBase {
            config: config.unwrap_or_default(),
            description: None,
            enumeration: self.enumeration,
            injection: self.injection,
            policies: self.policies,
            runtime: self.runtime_idx,
            title: self.name,
            as_id: self.as_id,
        })
    }

    pub fn enum_<T: std::fmt::Display>(mut self, enumeration: Option<&[T]>) -> Self {
        self.enumeration =
            enumeration.map(|enumeration| enumeration.iter().map(|v| format!("{v}")).collect());
        self
    }

    // TODO use a wit generated injection type; the this will be a in-place mutation
    fn convert_injection(&mut self, mut injection: Injection) -> Result<Injection> {
        if let Injection::Parent(data) = &mut injection {
            for type_id in data.values_mut().into_iter() {
                let type_def = TypeId(*type_id).resolve_ref()?.1;
                *type_id = self
                    .ctx
                    .register_type(type_def, Some(self.runtime_idx))?
                    .into();
            }
        }

        Ok(injection)
    }

    pub fn inject(mut self, injection: Option<Box<Injection>>) -> Result<Self> {
        self.injection = injection.map(|i| self.convert_injection(*i)).transpose()?;
        Ok(self)
    }

    pub fn id(mut self, b: bool) -> Self {
        self.as_id = b;
        self
    }
}
