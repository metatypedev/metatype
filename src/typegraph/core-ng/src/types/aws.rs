#[derive(Debug, Clone)]
pub struct S3RuntimeData {
    pub host_secret: String,
    pub region_secret: String,
    pub access_key_secret: String,
    pub secret_key_secret: String,
    pub path_style_secret: String,
}

#[derive(Debug, Clone)]
pub struct S3PresignGetParams {
    pub bucket: String,
    pub expiry_secs: Option<u32>,
}

#[derive(Debug, Clone)]
pub struct S3PresignPutParams {
    pub bucket: String,
    pub expiry_secs: Option<u32>,
    pub content_type: Option<String>,
}
