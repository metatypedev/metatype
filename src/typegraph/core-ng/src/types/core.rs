#[derive(Debug, Clone)]
pub struct Cors {
    pub allow_origin: Vec<String>,
    pub allow_headers: Vec<String>,
    pub expose_headers: Vec<String>,
    pub allow_methods: Vec<String>,
    pub allow_credentials: bool,
    pub max_age_sec: Option<u32>,
}

#[derive(Debug, Clone)]
pub struct Rate {
    pub window_limit: u32,
    pub window_sec: u32,
    pub query_limit: u32,
    pub context_identifier: Option<String>,
    pub local_excess: u32,
}

#[derive(Debug, Clone)]
pub struct TypegraphInitParams {
    pub name: String,
    pub dynamic: Option<bool>,
    pub path: String,
    // TypeMeta
    pub prefix: Option<String>,
    pub cors: Cors,
    pub rate: Option<Rate>,
}

#[derive(Debug, Clone)]
pub struct Artifact {
    pub path: String,
    pub hash: String,
    pub size: u32,
}

#[derive(Debug, Clone, Copy)]
pub struct MigrationAction {
    pub apply: bool,
    pub create: bool,
    pub reset: bool,
}

#[derive(Debug, Clone)]
pub struct PrismaMigrationConfig {
    pub migrations_dir: String,
    pub migration_actions: Vec<(String, MigrationAction)>,
    pub default_migration_action: MigrationAction,
}

#[derive(Debug, Clone)]
pub struct SerializeParams {
    pub typegraph_path: String,
    pub prefix: Option<String>,
    pub artifact_resolution: bool,
    pub codegen: bool,
    pub prisma_migration: PrismaMigrationConfig,
    pub pretty: bool,
}

pub type TypeId = u32;

#[derive(Debug, Clone, Default)]
pub struct TypeBase {
    pub name: Option<String>,
    // String => json String
    pub runtime_config: Option<Vec<(String, String)>>,
}

#[derive(Debug, Clone)]
pub struct TypeProxy {
    pub name: String,
    pub extras: Vec<(String, String)>,
}

#[derive(Debug, Clone, Default)]
pub struct TypeInteger {
    pub min: Option<i32>,
    pub max: Option<i32>,
    pub exclusive_minimum: Option<i32>,
    pub exclusive_maximum: Option<i32>,
    pub multiple_of: Option<i32>,
    pub enumeration: Option<Vec<i32>>,
}

#[derive(Debug, Clone, Default)]
pub struct TypeFloat {
    pub min: Option<f64>,
    pub max: Option<f64>,
    pub exclusive_minimum: Option<f64>,
    pub exclusive_maximum: Option<f64>,
    pub multiple_of: Option<f64>,
    pub enumeration: Option<Vec<f64>>,
}

#[derive(Debug, Clone, Default)]
pub struct TypeString {
    pub min: Option<u32>,
    pub max: Option<u32>,
    pub format: Option<String>,
    pub pattern: Option<String>,
    pub enumeration: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default)]
pub struct TypeFile {
    pub min: Option<u32>,
    pub max: Option<u32>,
    pub allow: Option<Vec<String>>,
}

#[derive(Debug, Clone)]
pub struct TypeList {
    pub of: TypeId,
    pub min: Option<u32>,
    pub max: Option<u32>,
    pub unique_items: Option<bool>,
}

#[derive(Debug, Clone)]
pub struct TypeOptional {
    pub of: TypeId,
    pub default_item: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct TypeUnion {
    pub variants: Vec<TypeId>,
}

#[derive(Debug, Clone, Default)]
pub struct TypeEither {
    pub variants: Vec<TypeId>,
}

#[derive(Debug, Clone, Default)]
pub struct TypeStruct {
    pub props: Vec<(String, TypeId)>,
    pub additional_props: bool,
    pub min: Option<u32>,
    pub max: Option<u32>,
    pub enumeration: Option<Vec<String>>,
}

#[derive(Debug, Clone)]
pub enum ValueSource {
    Raw(String),     // json
    Context(String), // key
    Secret(String),  // key
    Parent(String),  // name
    Param(String),   // name
}

#[derive(Debug, Clone)]
pub struct ParameterTransform {
    pub resolver_input: TypeId,
    pub transform_tree: String,
}

#[derive(Debug, Clone)]
pub struct TypeFunc {
    pub inp: TypeId,
    pub parameter_transform: Option<ParameterTransform>,
    pub out: TypeId,
    pub mat: MaterializerId,
    pub rate_calls: bool,
    pub rate_weight: Option<u32>,
}

#[derive(Debug, Clone)]
pub struct TransformData {
    pub query_input: TypeId,
    pub parameter_transform: ParameterTransform,
}

pub type PolicyId = u32;

#[derive(Debug, Clone)]
pub struct Policy {
    pub name: String,
    pub materializer: MaterializerId,
}

#[derive(Debug, Clone)]
pub struct PolicyPerEffect {
    pub read: Option<PolicyId>,
    pub create: Option<PolicyId>,
    pub update: Option<PolicyId>,
    pub delete: Option<PolicyId>,
}

#[derive(Debug, Clone)]
pub enum PolicySpec {
    Simple(PolicyId),
    PerEffect(PolicyPerEffect),
}

#[derive(Debug, Clone)]
pub enum ContextCheck {
    NotNull,
    Value(String),
    Pattern(String),
}

pub type RuntimeId = u32;
pub type MaterializerId = u32;

#[derive(Debug, Clone)]
pub struct FuncParams {
    pub inp: TypeId,
    pub out: TypeId,
    pub mat: MaterializerId,
}
