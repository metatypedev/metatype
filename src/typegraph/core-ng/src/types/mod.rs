// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

#[allow(unused)]
pub mod aws;
pub mod builders;
#[allow(unused)]
pub mod core;
#[allow(unused)]
pub mod runtimes;
pub mod type_def;
pub mod type_id;
pub mod type_ref;
#[allow(unused)]
pub mod utils;

use std::borrow::Cow;

pub use type_def::*;
pub use type_id::*;
pub use type_ref::*;

use crate::{errors::Result, types::core::TypeId as CoreTypeId};

#[derive(Clone, Debug)]
pub enum Type {
    Ref(TypeRef),
    Def(TypeDef),
}

impl From<TypeRef> for Type {
    fn from(r: TypeRef) -> Self {
        Self::Ref(r)
    }
}

impl From<TypeDef> for Type {
    fn from(d: TypeDef) -> Self {
        Self::Def(d)
    }
}

impl Type {
    fn name(&self) -> Option<Cow<'_, str>> {
        match self {
            Type::Ref(typ) => match typ.flatten().target {
                FlatTypeRefTarget::Direct(type_def) => {
                    type_def.name().map(ToString::to_string).map(Into::into)
                }
                FlatTypeRefTarget::Indirect(name) => Some(name.into()),
            },
            Type::Def(typ) => typ.name().map(Into::into),
        }
    }

    fn repr(&self) -> String {
        match self {
            Type::Ref(typ) => typ.repr(),
            Type::Def(typ) => typ.repr(),
        }
    }
}

pub fn get_type_repr(type_id: CoreTypeId) -> Result<String> {
    TypeId(type_id).repr()
}
