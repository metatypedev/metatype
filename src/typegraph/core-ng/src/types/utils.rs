#[derive(Debug, Clone)]
pub struct ReduceValue {
    pub inherit: bool,
    // json String
    pub payload: Option<String>,
}

#[derive(Debug, Clone)]
pub struct ReducePath {
    pub path: Vec<String>,
    pub value: ReduceValue,
}

#[derive(Debug, Clone)]
pub struct Reduce {
    pub paths: Vec<ReducePath>,
}

#[derive(Debug, Clone)]
pub enum AuthProtocol {
    Oauth2,
    Jwt,
    Basic,
}

#[derive(Debug, Clone)]
pub struct Auth {
    pub name: String,
    pub protocol: AuthProtocol,
    // String => json String
    pub auth_data: Vec<(String, String)>,
}

#[derive(Debug, Clone)]
pub struct QueryDeployParams {
    pub tg: String,
    pub secrets: Option<Vec<(String, String)>>,
}

#[derive(Debug, Clone)]
pub struct FdkConfig {
    pub workspace_path: String,
    pub target_name: String,
    pub config_json: String,
    pub tg_json: String,
}

#[derive(Debug, Clone)]
pub struct FdkOutput {
    pub path: String,
    pub content: String,
    pub overwrite: bool,
}
