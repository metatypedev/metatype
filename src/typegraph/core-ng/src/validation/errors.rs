// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::TgError;

// pub fn invalid_runtime_type(runtime: &str, materializer: &str) -> Error {
//     format!(
//         "invalid runtime type {} for materializer {}",
//         runtime, materializer
//     )
// }

pub fn invalid_output_type_predefined(name: &str, expected: &str, got: &str) -> TgError {
    format!(
        "invalid output type for predefined function {}: expected {}, got {}",
        name, expected, got
    )
    .into()
}

pub fn unknown_predefined_function(name: &str) -> TgError {
    format!("unknown predefined function {}", name).into()
}
