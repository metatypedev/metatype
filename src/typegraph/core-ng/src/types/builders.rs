// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use std::collections::HashSet;

use crate::{
    errors::{self, Result},
    global_store::{NameRegistration, Store},
};

use super::core::{
    TypeBase, TypeEither, TypeFile, TypeFloat, TypeFunc, TypeId as CoreTypeId, TypeInteger,
    TypeList, TypeOptional, TypeString, TypeStruct, TypeUnion,
};

use super::{
    AsId, Boolean, Either, File, Float, Func, Integer, List, Optional, StringT, Struct,
    TypeBoolean, TypeDef, TypeId, TypeRef, Union,
};

pub fn refb(name: String, attr: Option<String>) -> Result<CoreTypeId> {
    Ok(TypeRef::indirect(
        name,
        attr.map(|attr| {
            serde_json::from_str(&attr)
                .map_err(|e| format!("Could not parse ref attributes: {e:?}"))
        })
        .transpose()?,
    )?
    .id
    .0)
}

pub fn integerb(data: TypeInteger, base: TypeBase) -> Result<CoreTypeId> {
    if let (Some(min), Some(max)) = (data.min, data.max) {
        if min >= max {
            return Err(errors::invalid_max_value());
        }
    }
    if let (Some(min), Some(max)) = (data.exclusive_minimum, data.exclusive_maximum) {
        if min >= max {
            return Err(errors::invalid_max_value());
        }
    }
    Ok(Store::register_type_def(
        |id| TypeDef::Integer(Integer { id, base, data }.into()),
        NameRegistration(true),
    )?
    .into())
}

pub fn floatb(data: TypeFloat, base: TypeBase) -> Result<CoreTypeId> {
    if let (Some(min), Some(max)) = (data.min, data.max) {
        if min >= max {
            return Err(errors::invalid_max_value());
        }
    }
    if let (Some(min), Some(max)) = (data.exclusive_minimum, data.exclusive_maximum) {
        if min >= max {
            return Err(errors::invalid_max_value());
        }
    }
    Ok(Store::register_type_def(
        |id| TypeDef::Float(Float { id, base, data }.into()),
        NameRegistration(true),
    )?
    .into())
}

pub fn booleanb(base: TypeBase) -> Result<CoreTypeId> {
    Ok(Store::register_type_def(
        |id| {
            TypeDef::Boolean(
                Boolean {
                    id,
                    base,
                    data: TypeBoolean,
                }
                .into(),
            )
        },
        NameRegistration(true),
    )?
    .into())
}

pub fn stringb(data: TypeString, base: TypeBase) -> Result<CoreTypeId> {
    if let (Some(min), Some(max)) = (data.min, data.max) {
        if min >= max {
            return Err(errors::invalid_max_value());
        }
    }
    Ok(Store::register_type_def(
        |id| TypeDef::String(StringT { id, base, data }.into()),
        NameRegistration(true),
    )?
    .into())
}

pub fn as_id(type_id: CoreTypeId, composite: bool) -> Result<CoreTypeId> {
    TypeId(type_id).as_type()?.as_id(composite).map(|t| t.id.0)
}

pub fn fileb(data: TypeFile, base: TypeBase) -> Result<CoreTypeId> {
    if let (Some(min), Some(max)) = (data.min, data.max) {
        if min >= max {
            return Err(errors::invalid_max_value());
        }
    }
    Ok(Store::register_type_def(
        |id| {
            /* // TODO why??
            let base = TypeBase {
                name: Some(format!("_{}_file", id.0)),
                ..base
            }; */
            TypeDef::File(File { id, base, data }.into())
        },
        NameRegistration(true),
    )?
    .into())
}

pub fn listb(data: TypeList, base: TypeBase) -> Result<CoreTypeId> {
    if let (Some(min), Some(max)) = (data.min, data.max) {
        if min > max {
            return Err(errors::invalid_max_value());
        }
    }
    Ok(Store::register_type_def(
        |id| TypeDef::List(List { id, base, data }.into()),
        NameRegistration(true),
    )?
    .into())
}

pub fn optionalb(data: TypeOptional, base: TypeBase) -> Result<CoreTypeId> {
    /* let inner_name = match base.name {
        Some(_) => None,
        None => TypeId(data.of).name()?,
    }; */
    Ok(Store::register_type_def(
        |id| {
            /* let base = match inner_name {
                Some(n) => TypeBase {
                    name: Some(format!("_{}_{}?", id.0, n)),
                    ..base
                },
                None => base,
            }; */
            TypeDef::Optional(Optional { id, base, data }.into())
        },
        NameRegistration(true),
    )?
    .into())
}

pub fn unionb(data: TypeUnion, base: TypeBase) -> Result<CoreTypeId> {
    Ok(Store::register_type_def(
        |id| TypeDef::Union(Union { id, base, data }.into()),
        NameRegistration(true),
    )?
    .into())
}

pub fn eitherb(data: TypeEither, base: TypeBase) -> Result<CoreTypeId> {
    Ok(Store::register_type_def(
        |id| TypeDef::Either(Either { id, base, data }.into()),
        NameRegistration(true),
    )?
    .into())
}

pub fn structb(data: TypeStruct, base: TypeBase) -> Result<CoreTypeId> {
    let mut prop_names = HashSet::new();
    for (name, _) in data.props.iter() {
        if prop_names.contains(name) {
            return Err(errors::duplicate_key(name));
        }
        prop_names.insert(name.clone());
    }

    Ok(Store::register_type_def(
        |id| TypeDef::Struct(Struct { id, base, data }.into()),
        NameRegistration(true),
    )?
    .into())
}

pub fn funcb(data: TypeFunc) -> Result<CoreTypeId> {
    let wrapper_type = TypeId(data.inp);
    if !matches!(TypeDef::try_from(wrapper_type)?, TypeDef::Struct(_)) {
        return Err(errors::invalid_input_type(&wrapper_type.repr()?));
    }

    let base = TypeBase::default();
    Ok(Store::register_type_def(
        |id| TypeDef::Func(Func { id, base, data }.into()),
        NameRegistration(true),
    )?
    .into())
}
