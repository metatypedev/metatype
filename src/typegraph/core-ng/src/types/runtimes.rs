use super::core::{RuntimeId, TypeId};

pub type Idempotency = bool;

#[derive(Debug, Clone)]
pub enum Effect {
    Read,
    Create(Idempotency),
    Update(Idempotency),
    Delete(Idempotency),
}

#[derive(Debug, Clone)]
pub struct BaseMaterializer {
    pub runtime: RuntimeId,
    pub effect: Effect,
}

// deno
#[derive(Debug, Clone)]
pub struct MaterializerDenoFunc {
    pub code: String,
    pub secrets: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct MaterializerDenoStatic {
    pub value: String,
}

#[derive(Debug, Clone)]
pub struct MaterializerDenoPredefined {
    pub name: String,
}

#[derive(Debug, Clone)]
pub struct MaterializerDenoImport {
    pub func_name: String,
    pub module: String,
    pub deps: Vec<String>,
    pub secrets: Vec<String>,
}

// graphql
#[derive(Debug, Clone)]
pub struct GraphqlRuntimeData {
    pub endpoint: String,
}

#[derive(Debug, Clone)]
pub struct MaterializerGraphqlQuery {
    pub path: Option<Vec<String>>,
}

#[derive(Debug, Clone)]
pub struct HttpRuntimeData {
    pub endpoint: String,
    pub cert_secret: Option<String>,
    pub basic_auth_secret: Option<String>,
}

#[derive(Debug, Clone, Copy)]
pub enum HttpMethod {
    Get,
    Post,
    Put,
    Patch,
    Delete,
}

#[derive(Debug, Clone)]
pub struct MaterializerHttpRequest {
    pub method: HttpMethod,
    pub path: String,
    pub content_type: Option<String>,
    pub header_prefix: Option<String>,
    pub query_fields: Option<Vec<String>>,
    pub rename_fields: Option<Vec<(String, String)>>,
    pub body_fields: Option<Vec<String>>,
    pub auth_token_field: Option<String>,
}

// python
#[derive(Debug, Clone)]
pub struct MaterializerPythonDef {
    pub runtime: RuntimeId,
    pub name: String,
    pub fn_: String,
}

#[derive(Debug, Clone)]
pub struct MaterializerPythonLambda {
    pub runtime: RuntimeId,
    pub fn_: String,
}

#[derive(Debug, Clone)]
pub struct MaterializerPythonModule {
    pub runtime: RuntimeId,
    pub file: String,
    pub deps: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct MaterializerPythonImport {
    pub module: u32,
    pub func_name: String,
    pub secrets: Vec<String>,
}

// TODO: host:port

// random
#[derive(Debug, Clone)]
pub struct RandomRuntimeData {
    pub seed: Option<u32>,
    pub reset: Option<String>,
}

#[derive(Debug, Clone)]
pub struct MaterializerRandom {
    pub runtime: RuntimeId,
}

// wasm
#[derive(Debug, Clone)]
pub struct WasmRuntimeData {
    pub wasm_artifact: String,
}

#[derive(Debug, Clone)]
pub struct MaterializerWasmReflectedFunc {
    pub func_name: String,
}

#[derive(Debug, Clone)]
pub struct MaterializerWasmWireHandler {
    pub func_name: String,
}

// prisma
#[derive(Debug, Clone)]
pub struct PrismaRuntimeData {
    pub name: String,
    pub connection_string_secret: String,
}

#[derive(Debug, Clone)]
pub struct PrismaLinkData {
    pub target_type: TypeId,
    pub relationship_name: Option<String>,
    pub foreign_key: Option<bool>,
    pub target_field: Option<String>,
    pub unique: Option<bool>,
}

// prisma_migrate
#[derive(Debug, Clone)]
pub enum PrismaMigrationOperation {
    Diff,
    Create,
    Apply,
    Deploy,
    Reset,
}

// temporal
#[derive(Debug, Clone)]
pub struct TemporalRuntimeData {
    pub name: String,
    pub host_secret: String,
    pub namespace_secret: Option<String>,
}

#[derive(Debug, Clone)]
pub enum TemporalOperationType {
    StartWorkflow,
    SignalWorkflow,
    QueryWorkflow,
    DescribeWorkflow,
}

#[derive(Debug, Clone)]
pub struct TemporalOperationData {
    pub mat_arg: Option<String>,
    pub func_arg: Option<TypeId>,
    pub func_out: Option<TypeId>,
    pub operation: TemporalOperationType,
}

// typegate
pub enum TypegateOperation {
    VecTypegraphs,
    FindTypegraph,
    AddTypegraph,
    RemoveTypegraphs,
    GetSerializedTypegraph,
    GetArgInfoByPath,
    FindAvailableOperations,
    FindPrismaModels,
    RawPrismaRead,
    RawPrismaCreate,
    RawPrismaUpdate,
    RawPrismaDelete,
    QueryPrismaModel,
}

// typegraph  (introspection)
#[derive(Debug, Clone)]
pub enum TypegraphOperation {
    Resolver,
    GetType,
    GetSchema,
}

// substantial
#[derive(Debug, Clone)]
pub struct RedisBackend {
    pub connection_string_secret: String,
}

#[derive(Debug, Clone)]
pub enum SubstantialBackend {
    Memory,
    Fs,
    Redis(RedisBackend),
}

#[derive(Debug, Clone)]
pub struct SubstantialRuntimeData {
    pub backend: SubstantialBackend,
}

#[derive(Debug, Clone)]
pub enum WorkflowKind {
    Python,
    Deno,
}

#[derive(Debug, Clone)]
pub struct Workflow {
    pub name: String,
    pub file: String,
    pub deps: Vec<String>,
    pub kind: WorkflowKind,
}

#[derive(Debug, Clone)]
pub enum SubstantialOperationType {
    Start(Workflow),
    Stop(Workflow),
    Send(Workflow),
    Resources(Workflow),
    Results(Workflow),
}

#[derive(Debug, Clone)]
pub struct SubstantialOperationData {
    pub func_arg: Option<TypeId>,
    pub func_out: Option<TypeId>,
    pub operation: SubstantialOperationType,
}

// kv
#[derive(Debug, Clone)]
pub struct KvRuntimeData {
    pub url: String,
}

#[derive(Debug, Clone)]
pub enum KvMaterializer {
    Get,
    Set,
    Delete,
    Keys,
    Values,
}

// Grpc
#[derive(Debug, Clone)]
pub struct GrpcRuntimeData {
    pub proto_file: String,
    pub endpoint: String,
}

#[derive(Debug, Clone)]
pub struct GrpcData {
    pub method: String,
}
