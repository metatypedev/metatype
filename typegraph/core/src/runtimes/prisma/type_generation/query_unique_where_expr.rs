// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::types::TypeFun;
use crate::{global_store::with_store, types::TypeId};

use super::{TypeGen, TypeGenContext};

pub struct QueryUniqueWhereExpr {
    model_id: TypeId,
}

impl QueryUniqueWhereExpr {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for QueryUniqueWhereExpr {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let props = with_store(|s| -> Result<_> {
            let model = s.type_as_struct(self.model_id).unwrap();

            let mut props = vec![];
            for (k, ty) in model.data.props.iter() {
                let attrs = s.get_attributes(ty.into())?;
                let is_id = attrs
                    .concrete_type
                    .as_type(s)?
                    .get_base()
                    .ok_or_else(|| "expected a concrete type".to_string())?
                    .as_id;
                let is_unique = s.get_type(ty.into())?.get_base().map_or(false, |base| {
                    base.runtime_config
                        .iter()
                        .flatten()
                        .find_map(|(k, v)| (k == "unique").then(|| v.clone()))
                        .map_or(false, |v| v == "true")
                });
                if s.is_func(attrs.concrete_type)? || (!is_id && !is_unique) {
                    continue;
                }
                props.push((k.clone(), s.resolve_quant(attrs.concrete_type)?));
            }

            Ok(props)
        })?;

        let mut st = t::struct_();
        for (k, ty) in props.into_iter() {
            st.prop(k, t::optional(ty).build()?);
        }
        st.named(self.name(context)).build()
    }

    fn name(&self, _context: &TypeGenContext) -> String {
        let name = with_store(|s| s.get_type_name(self.model_id).unwrap().unwrap().to_string());
        format!("QueryUnique{}WhereInput", name)
    }
}
