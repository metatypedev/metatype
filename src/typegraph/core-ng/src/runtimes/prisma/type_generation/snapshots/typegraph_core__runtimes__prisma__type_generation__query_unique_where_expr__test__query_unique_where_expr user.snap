---
source: typegraph/core/src/runtimes/prisma/type_generation/query_unique_where_expr.rs
expression: "tree::print(ty)"
---
root: struct 'QueryUniqueUserWhereInput' #9
└─ [id]: optional #8
   └─ item: integer #0

